File: python/tests/test_centres_chunking.py
"""v2.2.x — single-multiset centres-path n_q chunking regression test.

Catches the OOM that surfaced when demo_triadConsonance ran at the
10-cent grid: K=72, r=3, rel, non-periodic, n_q=29161 — the
difference tensor would have been 2 × 357840 × 29161 = 155 GB
unchunked.

This regression slipped through because every existing routing /
parity case used n_q in the 20-50 range, well below the chunking
threshold. Any new fast-path or routing change that touches
centres-path evaluation MUST keep this test passing — verifying
small-n_q parity alone is not sufficient.
"""
from __future__ import annotations

import numpy as np
import pytest

import mpt
from mpt import build_exp_tens, eval_exp_tens


class TestCentresChunking:
    """Regression tests for memory-aware n_q chunking in the single-multiset
    centres fast-path.
    """

    def test_completes_without_oom_at_demo_scale(self):
        """K=72, r=3, rel, non-per at n_q=300 triggers chunking
        under the 1 GB per-chunk budget (per-value ~8.6 MB →
        2.6 GB unchunked). Should complete and return finite output.
        """
        K = 72
        p = np.linspace(0, 1200, K, endpoint=False)
        dens = build_exp_tens(p, np.ones(K), 12.0, 3,
                              True, False, 0.0, verbose=False)
        rng = np.random.default_rng(42)
        x = rng.uniform(0, 1200, (2, 300))

        v = eval_exp_tens(dens, x, method='centres', verbose=False)
        assert v.shape == (300,)
        assert np.all(np.isfinite(v))

    def test_chunked_matches_manual_split(self):
        """Chunking is purely a memory-management detail: per-query
        output depends only on the q-th query, so the full chunked
        call must equal the concatenation of per-half results to
        numerical precision. Matmul block-size differences between
        chunks of different sizes can introduce ULP-level variation,
        so we use a tight relative tolerance rather than bit-identity.
        """
        K = 72
        p = np.linspace(0, 1200, K, endpoint=False)
        dens = build_exp_tens(p, np.ones(K), 12.0, 3,
                              True, False, 0.0, verbose=False)
        rng = np.random.default_rng(42)
        x = rng.uniform(0, 1200, (2, 300))

        v_full = eval_exp_tens(dens, x, method='centres', verbose=False)

        mid = x.shape[1] // 2
        v_half1 = eval_exp_tens(dens, x[:, :mid],
                                method='centres', verbose=False)
        v_half2 = eval_exp_tens(dens, x[:, mid:],
                                method='centres', verbose=False)
        v_manual = np.concatenate([v_half1, v_half2])

        assert v_full.shape == v_manual.shape == (300,)
        np.testing.assert_allclose(v_full, v_manual, rtol=1e-13, atol=0)

    def test_auto_routing_matches_forced_centres(self):
        """For K=48 r=3 rel PERIODIC at sigma/P = 1/120, auto routing
        must produce the same values as forced 'centres' to within the
        accuracy floor.

        Auto is free to route to whichever path the cost model prices
        cheaper (for this shape the Möbius path is measured several
        times faster than centres, so auto legitimately returns Möbius
        values); the invariant is routing-independent correctness, not
        path identity. Möbius values agree with centres only to the
        u-grid quadrature accuracy tied to truncation_sigmas (the suite
        baseline resolves to the 1e-12 floor), so the comparison is a
        floor-scaled allclose rather than bitwise.
        """
        K = 48
        p = np.linspace(0, 1200, K, endpoint=False)
        dens = build_exp_tens(p, np.ones(K), 10.0, 3,
                              True, True, 1200.0, verbose=False)
        rng = np.random.default_rng(42)
        x = rng.uniform(0, 1200, (2, 300))

        # Pin kernel_chunk_bytes to a fixed value so both paths compute
        # their chunk sizes against the same byte budget. The function-
        # scoped defaults-reset fixture in conftest.py restores the
        # factory state on teardown.
        mpt.set_default(kernel_chunk_bytes=200_000_000)

        v_auto = eval_exp_tens(dens, x, verbose=False)
        v_centres = eval_exp_tens(dens, x, method='centres', verbose=False)

        peak = float(np.max(np.abs(v_centres)))
        np.testing.assert_allclose(v_auto, v_centres,
                                   rtol=0.0, atol=1e-10 * peak)

File: matlab/tests/test_centres_chunking.m
%% test_centres_chunking.m — v3 single multiset centres-path n_q chunking
%
%  Regression test. Catches the OOM that surfaced when
%  demo_triadConsonance ran at the 10-cent grid: K=72, r=3, rel,
%  non-periodic, n_q=29161 — the difference tensor would have been
%  2 x 357840 x 29161 = 155 GB unchunked, exceeding MATLAB's
%  default 48 GB array-size cap.
%
%  This regression slipped through the original test suite because
%  every existing routing/parity case used n_q in the 20-50 range,
%  well below the chunking threshold for any reasonable density.
%  Any new fast-path or routing change that touches centres-path
%  evaluation MUST keep this test passing — verifying small-n_q
%  parity alone is not sufficient.
%
%  The test strategy:
%   (a) Build the same density as the demo (K=72, r=3, rel).
%   (b) Call evalExpTens at n_q=1000 — large enough to overflow the
%       difference-tensor allocation under any chunking budget below
%       ~5 GB, but small enough for a routine test run (~few seconds).
%   (c) Verify (i) no OOM, (ii) all outputs finite, (iii) the chunked
%       result equals the concatenation of two manual halves
%       (FP-bit-identical, since per-query output depends only on
%       the q-th query — chunking only changes memory footprint).
%
%  Standalone-runnable. When invoked from test_mpt.m the existing
%  `results` cell is appended to.

if ~exist('results', 'var')
    results = {};
    % Defaults isolation when run standalone (when invoked from
    % test_mpt.m the outer wrapper has already isolated defaults).
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults_cc
    cleanupDefaults_cc = mptTestIsolateDefaults(); %#ok<NASGU>
end

% --- Density: matches demo_triadConsonance ---
%   nJ = 72 * 71 * 70 = 357,840
%   dim = r - 1 = 2 (rel mode)
%   Per-n_q memory in the difference tensor:
%     (dim + 1) * nJ * 8 = 8.59 MB
%   At n_q = 1000: 8.59 GB unchunked — triggers chunking under
%   MATLAB's default budget (memInfo.MaxPossibleArrayBytes * 0.5)
%   on systems with < 17 GB free, and certainly under the 4 GB
%   fallback used when `memory` is unavailable.
cc_K = 72;
cc_p = linspace(0, 1200, cc_K + 1)';
cc_p = cc_p(1:end-1);
cc_w = ones(cc_K, 1);
cc_dens = buildExpTens(cc_p, cc_w, 12, 3, true, false, 1200, ...
    'verbose', false);

rng(42, 'twister');
cc_X = rand(2, 1000) * 1200;

% --- Full call: fast-path with internal n_q chunking ---
cc_vFull = evalExpTens(cc_dens, cc_X, 'method', 'centres', 'verbose', false);

% --- Manual split: two halves, each potentially chunked too ---
cc_n = size(cc_X, 2);
cc_mid = floor(cc_n / 2);
cc_v1 = evalExpTens(cc_dens, cc_X(:, 1:cc_mid), ...
    'method', 'centres', 'verbose', false);
cc_v2 = evalExpTens(cc_dens, cc_X(:, cc_mid + 1:end), ...
    'method', 'centres', 'verbose', false);
cc_vManual = [cc_v1, cc_v2];

% --- Assertions ---
results{end + 1, 1} = 'centres_chunking: completes without OOM (K=72, r=3, rel, n_q=1000)';
results{end, 2}     = numel(cc_vFull) == cc_n;

results{end + 1, 1} = 'centres_chunking: all outputs finite';
results{end, 2}     = all(isfinite(cc_vFull));

results{end + 1, 1} = 'centres_chunking: full call matches manual split (rel-tol 1e-13)';
results{end, 2}     = all(abs(cc_vFull - cc_vManual) <= 1e-13 * abs(cc_vFull));

% Note: the centres n_q chunking is exercised via method='centres' above,
% mirroring the Python chunking test (which forces centres throughout).
% Auto routing is deliberately not asserted here: the cost model routes
% this K=72 r=3 rel non-per shape to the factored Möbius path, so an auto
% call would not exercise the centres chunking path this test targets.
% Auto-vs-centres agreement is a dispatcher concern (test_dispatch_sm_eval,
% test_eval_orbit), not a chunking one.

clear cc_K cc_p cc_w cc_dens cc_X cc_vFull cc_n cc_mid cc_v1 cc_v2 ...
      cc_vManual

% Restore caller's pre-test defaults eagerly when run
% standalone (fires the helper's onCleanup destructor on
% script exit; guarded so we don't clear a like-named
% variable when this file was run from test_mpt.m, where
% the standalone branch was skipped).
if exist('cleanupDefaults_cc', 'var')
    clear cleanupDefaults_cc
end
